pkg/reconciler/revision/resources/pa.go

```diff
@@ -45,20 +46,27 @@ func MakePA(rev *v1.Revision) *autoscalingv1alpha1.PodAutoscaler {
 				Name:       names.Deployment(rev),
 			},
 			ProtocolType: rev.GetProtocol(),
-			Reachability: reachability(rev),
+			Reachability: reachability(rev, deployment),
 		},
 	}
 }
 
-func reachability(rev *v1.Revision) autoscalingv1alpha1.ReachabilityType {
+func reachability(rev *v1.Revision, deployment *appsv1.Deployment) autoscalingv1alpha1.ReachabilityType {
 	// check infra failures
-	conds := []apis.ConditionType{
+	infraFailure := false
+	for _, cond := range []apis.ConditionType{
 		v1.RevisionConditionResourcesAvailable,
 		v1.RevisionConditionContainerHealthy,
+	} {
+		if c := rev.Status.GetCondition(cond); c != nil && c.IsFalse() {
+			infraFailure = true
+			break
+		}
 	}
 
-	for _, cond := range conds {
-		if c := rev.Status.GetCondition(cond); c != nil && c.IsFalse() {
+	if infraFailure && deployment != nil && deployment.Spec.Replicas != nil {
+		// If we have an infra failure and no ready replicas - then this revision is unreachable
+		if *deployment.Spec.Replicas > 0 && deployment.Status.ReadyReplicas == 0 {
```

Comment:
So in the past we moved from checking if rev is active:
```
					if !rev.IsReachable() {
						return autoscalingv1alpha1.ReachabilityUnreachable
					}
``` 
to routing state but it was proven a bit too aggressive with the case of a broken webhook and the upgrade of a deployment, cutting traffic.
Could we do a follow up PR to document this state machine of resources, so we feel more confident about changing stuff?

Reply:
> Now with this patch we only set pa to "unreachable" due to revision being not healthy or active, only if there are no ready replicas (when >0 are required).

The other condition is if the revision is not being pointed to by a route - then it's unreachable as well.


> Btw what is the effect on the revision when we mark this unreachable, does it propagate to the revision?

If the revision marks the PA unreachable then the autoscaler will scale the deployment down to zero.

> Could we also do a follow up PR to document this state machine of resources, so we feel more confident about changing stuff?

Sure - I also included the necessary tests to cover this case